ext/standard/password.c

```diff
@@ -669,9 +669,9 @@ PHP_FUNCTION(password_hash)
 	algo = php_password_algo_find_zval(zalgo);
 	if (!algo) {
 		zend_string *algostr = zval_get_string(zalgo);
-		php_error_docref(NULL, E_WARNING, "Unknown password hashing algorithm: %s", ZSTR_VAL(algostr));
+		zend_value_error("Unknown password hashing algorithm: %s", ZSTR_VAL(algostr));
 		zend_string_release(algostr);
-		RETURN_NULL();
+		return;
 	}
 
 	digest = algo->hash(password, options);
```

Comment:
I'd use `ZEND_ASSERT(EG(exception)); return;` for this error condition (everything throws now, right?) and then it should be possible to remove `null` from the possible return values.

Reply:
Also, I really think we need to introduce a macro like `RETURN_THROWS()` or so that does this ... it would be nice to make clear which of the "return"s are expected to be due to an exception. But that's a separate change...

Reply:
I found three other cases where we don't currently throw, so I'll commit them now. As far as I can understand though, it won't be enough to remove `null` from return values. E.g. `php_password_bcrypt_hash()` still returns `NULL` in some cases without emitting a warning (https://github.com/php/php-src/blob/41673a527a48db22f099a55bb58a98b310ff90a8/ext/standard/password.c#L217). Or am I missing something?

Reply:
The one you referenced looks like a sanity check that is never expected to fail in practice. However, it does look like we shouldn't blindly assume that something has been thrown here. The alternative is to do `if (!EG(exception)) { zend_throw_error(NULL, "Password hashing failed for unknown reason"); }` as a catch-all if something is missed.

Reply:
Great idea! I've just pushed a new commit including this.